drivers/timer/cortex_m_systick.c

```diff
@@ -101,10 +101,14 @@ static uint32_t elapsed(void)
 	uint32_t ctrl = SysTick->CTRL;	/* B */
 	uint32_t val2 = SysTick->VAL;	/* C */
 
-	/* SysTick behavior: The counter wraps at zero automatically,
-	 * setting the COUNTFLAG field of the CTRL register when it
-	 * does.  Reading the control register automatically clears
-	 * that field.
+	/* SysTick behavior: The counter wraps after zero automatically.
+	 * The COUNTFLAG field of the CTRL register is set when it
+	 * decrements from 1 to 0. Reading the control register
```

Comment:
Is that really documented behavior?  There's really a one cycle race between the wraparound and the flag bit?  (i.e. it's possible in that cycle to have a flag set but the counter read "unexpired"?).    Do emulators like qemu have the same behavior?  I bet they don't, in which case this is going to introduce bugs on the CI platforms, no?

Reply:
>  Is that really documented behavior?
>  Do emulators like qemu have the same behavior?

Yes. A good way to understand this SysTick is to look at the policies that QEMU uses:

    s->ptimer = ptimer_init(systick_timer_tick, s,
                            PTIMER_POLICY_WRAP_AFTER_ONE_PERIOD |
                            PTIMER_POLICY_NO_COUNTER_ROUND_DOWN |
                            PTIMER_POLICY_NO_IMMEDIATE_RELOAD |
                            PTIMER_POLICY_TRIGGER_ONLY_ON_DECREMENT);

    /* Periodic timer counter stays with "0" for a one period before wrapping
     * around.  */
    #define PTIMER_POLICY_WRAP_AFTER_ONE_PERIOD (1 << 0)

    /* Make counter value of the running timer represent the actual value and
     * not the one less.  */
    #define PTIMER_POLICY_NO_COUNTER_ROUND_DOWN (1 << 4)

    /* Starting to run with/setting counter to "0" won't re-load counter
     * immediately, but after a one period.  */
    #define PTIMER_POLICY_NO_IMMEDIATE_RELOAD   (1 << 3)

    /*
     * Starting to run with a zero counter, or setting the counter to "0" via
     * ptimer_set_count() or ptimer_set_limit() will not trigger the timer
     * (though it will cause a reload). Only a counter decrement to "0"
     * will cause a trigger. Not compatible with NO_IMMEDIATE_TRIGGER;
     * ptimer_init() will assert() that you don't set both.
     */
    #define PTIMER_POLICY_TRIGGER_ONLY_ON_DECREMENT (1 << 5)
    
The last one "Only a counter decrement to "0" will cause a trigger" specifies that behavior.
(I have not dug up the actual reference manual yet... but I am pretty sure it will be in there).

Reply:
The [Armv7-m Architecture Reference Manual](https://developer.arm.com/documentation/ddi0403/ee/?lang=en) (chapter B3.3.1) says:
> When the counter transitions to zero, it sets the COUNTFLAG status bit to 1. Reading the COUNTFLAG status bit
clears it to 0.
>
>Writing to SYST_CVR clears both the register and the COUNTFLAG status bit to zero. This causes the SysTick logic to reload SYST_CVR from SYST_RVR on the next timer clock. A write to SYST_CVR does not trigger the SysTick exception logic.
>
> Reading SYST_CVR returns the value of the counter at the time the register is accessed.

where `SYST_CVR` is `SysTick->VAL` and `SYST_RVR` is `SysTick->LOAD`.

Armv6-m & Armv8-m are sensibly the same.
See:
- [Armv6-m Architecture Reference Manual](https://developer.arm.com/documentation/ddi0419/e/?lang=en) chapter B3.3.1
- [Armv8-m Architecture Reference Manual](https://developer.arm.com/documentation/ddi0553/bw/?lang=en) chapter B11.1